pkgs/build-support/php/build-php.nix

```diff
@@ -0,0 +1,138 @@
+{ stdenvNoCC, lib, writeTextDir, php, unzip, xz, git }@inputs:
+
+let
+  buildPhpProjectOverride = finalAttrs: previousAttrs:
+
+  let
+    composerLock = finalAttrs.composerLock or null;
+    vendorDir = finalAttrs.vendorDir or "vendor";
+    php = finalAttrs.php or inputs.php;
+
+    composer = finalAttrs.composer or inputs.php.packages.composer;
+
+    # See https://github.com/composer/composer/pull/9212
+    prefix = "${builtins.replaceStrings ["-"] [""] finalAttrs.pname}";
+
+    # See https://getcomposer.org/doc/03-cli.md#composer-home-config-json
+    composerHome = let
+      content = {
+        config = {
+          autoloader-suffix = prefix;
+        };
+      };
+    in
+      writeTextDir "/config.json" (builtins.toJSON content);
+
+    vendor = stdenvNoCC.mkDerivation (finalVendorAttrs: {
+      src = finalAttrs.src;
+      version = finalAttrs.src.rev;
+      pname = "${finalAttrs.pname}-vendor";
+
+      # See https://github.com/NixOS/nix/issues/6660
+      dontPatchShebangs = true;
+
+      nativeBuildInputs = [
+        composer
+        git
+        unzip
+        xz
+      ];
+
+      strictDeps = true;
+
+      configurePhase = ''
+        runHook preConfigure
+
+        ${lib.optionalString (composerLock != null) "cp ${composerLock} composer.lock"}
+
+        if [ ! -f "composer.lock" ]
+        then
+          echo "No composer.lock file found"
+          exit 1
+        fi
+
+        mkdir -p cache
+
+        runHook postConfigure
+      '';
+
+      buildPhase = ''
+        runHook preBuild
+
+        COMPOSER_CACHE_DIR=cache \
+        COMPOSER_HOME=${composerHome} \
+        composer install \
+          --download-only \
+          --no-dev \
+          --no-interaction \
+          --no-scripts
+
+        runHook postBuild
+      '';
+
+      installPhase = ''
+        runHook preInstall
+
+        mkdir -p $out
+        cp composer.lock $out/
+        cp -ar cache/ $out/
+
+        runHook postInstall
+      '';
+
+      outputHashMode = "recursive";
+      outputHashAlgo = if (finalAttrs ? vendorHash && finalAttrs.vendorHash != "") then null else "sha256";
+      outputHash = finalAttrs.vendorHash or "";
+    });
+  in {
+    name = finalAttrs.pname;
+
+    nativeBuildInputs = [
+      composer
+      git
+      unzip
+      xz
+    ];
+
+    buildInputs = [
+      php
+    ];
+
+    strictDeps = true;
+
+    configurePhase = ''
+      runHook preConfigure
+
+      cp -fr ${vendor}/* .
+      chmod 777 cache -R
+
+      runHook postConfigure
+    '';
+
+    buildPhase = ''
+      runHook preBuild
+
+      COMPOSER_DISABLE_NETWORK=1 \
+      COMPOSER_CACHE_DIR=cache \
+      COMPOSER_VENDOR_DIR=${vendorDir} \
```

Comment:
Why do we need to set vendor dir?

Reply:
To make sure it's not set in `composer.json` by the developer.

Reply:
But it does not look to me like we depend on it not being changed.

Reply:
True, since we're using `--download-only` this is now no more needed. I removed it.